rfcs/0041-new-transaction-model.md

```diff
@@ -0,0 +1,179 @@
+  assert (implied_root a merkle_path e.ledger_hash = e.ledger_hash) ;
+  assert (verify_authorization authorization a s.id) ;
+  assert (verify_predicate predicate a) ;
+  assert (auth_sufficient_given_permissions a.permissions authorization update) ;
+  let fee_excess_change =
+    match s.current_token_id with
+    | None -> a.token_id
+    | Some curr ->
+      if curr == a.token_id
+      then Currency.Fee.zero
+      else (
+        (* If we are switching tokens, then we cannot have printed money out of thin air. *)
+        assert (s.excess >= 0);
+        if curr == Token_id.default
+        then s.excess
+        else 0 (* burn the excess of this non-default token *)
+      )
+  in    
+  let a' = perform_update update a in
+  let excess = current_transaction_local_state.excess + (a.balance - a'.balance) in
+  let new_ledger_hash = implied_root a' merkle_path in
+  match remaining with
+  | [] ->
+  	assert (excess >= 0);
+  	let fee_excess_change =
+  	  fee_excess_change +
+  	  if a.token_id == Token_id.default then excess else 0
+  	in
+  	{ e with current_transaction_local_state= None
+  	; ledger_hash= new_ledger_hash
+  	; fee_excess= e.fee_excess + fee_excess_change }
+  | _::_ ->
+  	{ e with current_transaction_local_state=
+  	  Some
+  	    { local_state with parties= remaining
+  	    ; excess= local_state.excess + excess }
+  	; ledger_hash= new_ledger_hash 
+  	; fee_excess= e.fee_excess + excess_change }
+```
+
+### How this would boil down into "base" transaction SNARKs
+The idea would be to have 3 new base transaction SNARKs corresponding to the 3 forms of authentication. Each would implement the above `apply_step_or_start` function but with the `verify_authorization` specialized to either signature verification, SNARK verification, or nothing.
+
+
+Under this model, executing a transaction works as follows. Let `t = [t1; t2; t3]`
+
+### Fees and proofs required
+
+Instead of 2 proofs per transaction as is required now, we will switch to 2 proofs per "party".
+
+Similarly, we should switch to ordering transactions in the transaction pool by `fee / number of parties`.
+
+## Benefits of this approach
+
+The main benefit of this approach is that we can have a small number of base circuits, each of which has at most one verifier inside of it, while still supporting transactions containing arbitrary numbers of parties and proofs. This enables such applications as multi-token swaps and snapp interactions involving arbitrarily many accounts.
+
+Another benefit is the simplified and unified implementation for transaction application logic (both inside and outside the SNARK).
+
+## Eliminating all other transaction types
+
+Ideally, for simplicity, we will replace the implementation of transaction logic and transaction SNARK for the existing "special case" transactions (of payments and stake delegations) into sequences of "parties" as above. We can still keep the special case variants in the transaction type if desired.
```

Comment:
Can you add some discussion of how this will affect the number of proofs per transaction, whether we will need to batch some party proofs, and what this might do to costs? What you mentioned in our DMs would be a good starting point.